core/net/mac/tsch/sixtop/sixp.c

```diff
@@ -0,0 +1,967 @@
+}
+/*---------------------------------------------------------------------------*/
+static int
+transit_trans_state(sixp_trans_t *trans, sixp_trans_state_t new_state)
+{
+  assert(trans != NULL);
+
+  /* enforce state transition rules  */
+  if((new_state == SIXP_TRANS_STATE_REQUEST_SENT &&
+      trans->state != SIXP_TRANS_STATE_INIT) ||
+     (new_state == SIXP_TRANS_STATE_REQUEST_RECEIVED &&
+      trans->state != SIXP_TRANS_STATE_INIT) ||
+     (new_state == SIXP_TRANS_STATE_RESPONSE_SENT &&
+      trans->state != SIXP_TRANS_STATE_REQUEST_RECEIVED) ||
+     (new_state == SIXP_TRANS_STATE_RESPONSE_RECEIVED &&
+      trans->state != SIXP_TRANS_STATE_REQUEST_SENT) ||
+     (new_state == SIXP_TRANS_STATE_CONFIRMATION_SENT &&
+      trans->state != SIXP_TRANS_STATE_RESPONSE_RECEIVED &&
+      trans->mode == SIXP_TRANS_MODE_3_STEP) ||
+     (new_state == SIXP_TRANS_STATE_CONFIRMATION_RECEIVED &&
+      trans->state != SIXP_TRANS_STATE_RESPONSE_SENT &&
+      trans->mode == SIXP_TRANS_MODE_3_STEP)) {
+    /* invalid state transition */
+    return -1;
+  }
+
+  PRINTF("6top: trans %p state changes from %u to %u\n",
+         trans, trans->state, new_state);
+
+  trans->state = new_state;
+  schedule_trans_process(trans);
+
+  return 0;
+}
+/*---------------------------------------------------------------------------*/
+static sixp_trans_t *
+alloc_trans(sixp_code_t code, uint8_t sfid, uint8_t seqno,
+            const linkaddr_t *peer_addr, sixp_trans_mode_t mode)
+{
+  sixp_trans_t *trans;
+
+  assert(peer_addr != NULL);
+
+  if(sixtop_find_sf(sfid) == NULL) {
+    PRINTF("6top: cannot find the specified SF [sfid=%u]\n", sfid);
+    return NULL;
+  }
+
+  if(find_trans(peer_addr) != NULL) {
+    PRINTF("6top: trans with ");
+    PRINTLLADDR((const uip_lladdr_t *)peer_addr);
+    PRINTF(" already exists\n");
+    return NULL;
+  }
+
+  if((trans = memb_alloc(&trans_memb)) == NULL) {
+    PRINTF("6top: lack of memory\n");
+    return NULL;
+  }
+
+  memset(trans, 0, sizeof(sixp_trans_t));
+  trans->sf = sixtop_find_sf(sfid);
+  trans->peer_addr = *peer_addr;
+  trans->seqno = seqno;
+  trans->cmd = code.value;
+  trans->state = SIXP_TRANS_STATE_INIT;
+  trans->mode = mode;
+
+  list_add(trans_list, trans);
+
+  return trans;
+}
+/*---------------------------------------------------------------------------*/
+static void
+free_trans(sixp_trans_t *trans)
+{
+  assert(trans != NULL);
+
+  list_remove(trans_list, trans);
+  memb_free(&trans_memb, trans);
+}
+/*---------------------------------------------------------------------------*/
+static sixp_trans_t *
+find_trans(const linkaddr_t *peer_addr)
+{
+  sixp_trans_t *trans;
+
+  assert(peer_addr != NULL);
+
+  /*
+   * XXX: we don't support concurrent 6P transactions which is mentioned in
+   * Section 4.3.3, draft-ietf-6tisch-6top-protocol-03.
+   *
+   * The assumption here is that there is one transactions for a single peer at
+   * most.
```

Comment:
Sorry, I overlooked this comment somehow... [Section 4.3.3](https://tools.ietf.org/html/draft-ietf-6tisch-6top-protocol-03#section-4.3.3) says, "only a single 6P Transaction between two neighbors, in a given direction, can take place at the same time." Therefore, a node may have at most two transactions with the same peer, one of which is initiated by the node, the other is by the peer. The same thing is said by "nodes A and B MAY support having two transactions going on at the same time, one in each direction."